python/cuml/dask/common/utils.py

```diff
@@ -153,12 +160,106 @@ def patch_cupy_sparse_serialization(client):
 
     client : dask.distributed.Client client to use
     """
+
+    from distributed.protocol import register_generic
+
     def patch_func():
         def serialize_mat_descriptor(m):
             return cp.cupy.cusparse.MatDescriptor.create, ()
 
+        register_generic(Base)
+
         copyreg.pickle(cp.cupy.cusparse.MatDescriptor,
                        serialize_mat_descriptor)
 
     patch_func()
     client.run(patch_func)
+
+
+class MultiHolderLock:
+    """
+    A per-process synchronization lock allowing multiple concurrent holders
+    at any one time. This is used in situations where resources might be
+    limited and it's important that the number of concurrent users of
+    the resources are constained.
+
+    This lock is serializable, but relies on a Python threading.Lock
+    underneath to properly synchronize internal state across threads.
+    Note that this lock is only intended to be used per-process and
+    the underlying threading.Lock will not be serialized.
+    """
+
+    def __init__(self, n):
+        """
+        Initialize the lock
+        :param n : integer the maximum number of concurrent holders
+        """
+        self.n = n
+        self.current_tasks = 0
+        self.lock = Lock()
+
+    def _acquire(self, blocking=True, timeout=10):
+        lock_acquired = False
+        inner_lock_acquired = self.lock.acquire(blocking, timeout)
+
+        if inner_lock_acquired and self.current_tasks < self.n - 1:
+            self.current_tasks += 1
+            lock_acquired = True
+            self.lock.release()
+
+        return lock_acquired
+
+    def acquire(self, blocking=True, timeout=10):
+        """
+        Acquire the lock.
+        :param blocking : bool will block if True
+        :param timeout : a timeout (in seconds) to wait for the lock
+                         before failing.
+        :return : True if lock was acquired successfully, False otherwise
+        """
+
+        t = time.time()
+
+        lock_acquired = self._acquire(blocking, timeout)
+
+        while blocking and not lock_acquired:
+
+            if time.time() - t > timeout:
+                raise TimeoutError()
```

Comment:
Spinning here is a bit dangerous since it'll be hard for other threads to get running so they can release the lock. Adding a `time.sleep` with exponential backoff (but capped by timeout - time.time()) would be conventional I think.

Reply:
That is true, this should be sleeping here and sleep time should be variable.